forkrun.bash

```diff
@@ -1092,17 +1041,119 @@
             [[ -f "${tmpDir}"/.quit ]] && break
             source /proc/self/fd/0 <<<"${coprocSrcCode//'{<#>}'/"${kkProcs}"}"
         done
-        echo "${kkProcs}" >"${tmpDir}"/.nWorkers                    }
+        echo "${kkProcs}" >"${tmpDir}"/.nWorkers                    
         : >"${tmpDir}"/.spawned
 
         (( ${verboseLevel} > 1 )) && printf '\n\n%s WORKER COPROCS FORKED\n\n' "${nProcs}" >&${fd_stderr}
+
+        # setup dynamically coproc to spawn new workers based on read queue length
```

Comment:
1st suggestion:  Im trying to keep forkrun "self-contained" in a single file as much as possible. Also, sourcing the function to compute load when forkrun is sourced means it is already sourced when forkrun runs, helping to (slightly) reduce runtime.

2nd suggestion:  using the simpler spawning protocol doesnt work well in practice. The issue is that once a coproc is spawned it is active until the end...there is no "un-spawning" or removing it. So, we need to be more selective about when to spawn them, otherwise every time forkrun is run you will just end up spawning "$max_workers" worker coprocs...which sort of defeats the point of dynamically adjusting how many workers to spawn, since "always spawning $max_workers" isnt dynamic. 

3rd suggestion: for reasons related to "ensuring there arent any processes left running when the end condition is met and forkrun exits", it is (in my opinion) somewhat prefferable to spawn new coprocs directly from the forkrun function (or one of its coproc children) and not from another function.

Reply:
In a future update i might even add a stochastic element to spawning the coprocs. e.g., when all the existing conditions are met, get a random number and if it is above some cutoff spawn the coproc. The cutoff would change such that initially you were almost 100% guaranteed to get a random number above the cutoff, but as you get closer and closer to having "$max_workers" coprocs the chances of spawning another decrease more and more.

Its surprisingly tricky to get the number of additional workers dynamically spawned to not always end up at either 0 or at $max_workers...it is *really* hard to find criteria that stabalize this process so you end up somewhere in the middle (after potentially 10's of thousands of iterations, if not more) in a way that makes sense for dynamically adjusting the number of coprocs...